### PR TITLE
feat: add --server-name flag for configurable MCP server name

### LINKED ISSUE #965: Allow overriding the advertised MCP server name (--server-name flag / env var)
Is your feature request related to a problem?

When running multiple `mcp-grafana` instances against different Grafana instances (one container per Grafana), MCP clients can't tell them apart. The server advertises a hardcoded `serverInfo.name` of `mcp-grafana` during the initialize handshake (in `cmd/mcp-grafana/main.go`, currently `server.NewMCPServer("mcp-grafana", ...)`).

Clients that derive their tool namespace from this name — notably VS Code / GitHub Copilot — deduplicate identical names with numeric suffixes (`mcp-grafana`, `mcp-grafana1`, `mcp-grafana2`, …). The model only sees these meaningless suffixes, so a user asking to use a specific instance can't reliably target one. The suffix ordering also isn't stable across installs.

This is a real pain for anyone running several Grafana instances (e.g. per-project or per-environment monitoring) through a single MCP client.

Describe the solution you'd like

A way to override the advertised server name at startup — e.g. a `--server-name` CLI flag and/or a `GRAFANA_MCP_SERVER_NAME` environment variable, defaulting to `mcp-grafana` for full backward compatibility. Each instance could then be launched as `grafana-project-a`, `grafana-project-b`, etc., and clients would present them as distinct, nameable servers.

Describe alternatives you've considered

- Client-side workarounds (documenting which numeric suffix maps to which instance) — fragile and install-order dependent.
- Maintaining a fork that makes the name configurable — works (verified: a small change reading the flag/env var makes instances distinguishable in Copilot), but means maintaining a fork of a production dependency just for a label.

Additional context

The change is minimal — reading a flag/env var and passing it to `NewMCPServer` instead of the literal string. Happy to open a PR if you're open to it.

### PR DESCRIPTION
**Pull request by austinmilt**
## Summary

Closes https://github.com/grafana/mcp-grafana/issues/965

- Adds `--server-name` flag and `GRAFANA_MCP_SERVER_NAME` env var to let operators distinguish multiple mcp-grafana instances in the MCP handshake and OTel `service.name` resource attribute
- Validates names against `^[a-zA-Z0-9][a-zA-Z0-9._-]*$` (1–128 chars) to prevent log injection, path traversal, shell metacharacters, and metric cardinality issues
- Uses `flag.Visit` for reliable flag detection so `--server-name=mcp-grafana` correctly overrides the env var even when it matches the default
- Keeps instrumentation scope names (tracer, meter), HTTP User-Agent, and proxied-client name hardcoded to `"mcp-grafana"` per OTel conventions and security considerations
- Error messages identify the source (`--server-name` vs `GRAFANA_MCP_SERVER_NAME`) for actionable diagnostics

### Precedence

`--server-name` flag > `GRAFANA_MCP_SERVER_NAME` env var > default `"mcp-grafana"`

### What changes with a custom name

- MCP `initialize` response `serverInfo.name`
- OTel `service.name` resource attribute (may require updating monitoring dashboards)
- OTLP log bridge service identity

### What does NOT change

- OTel instrumentation scope names (tracer/meter) — these identify the library, not the instance
- HTTP `User-Agent` header — identifies the product, not the deployment
- Proxied MCP client name

## Test plan

- [x] `TestResolveServerName` — precedence rules including flag-explicitly-set-to-default edge case
- [x] `TestValidateServerName` — 28 cases covering valid names and attack vectors (injection, traversal, unicode, shell metacharacters)
- [x] `TestNewServer_DefaultServerName` — regression guard for default behavior
- [x] `TestNewServer_CustomServerName` — custom name flows through to MCP handshake
- [x] `TestNewServer_MultiInstanceDistinctNames` — two instances report distinct names
- [x] `TestCustomServerName_DoesNotAffectUserAgent` — custom name doesn't leak into HTTP headers
- [x] `TestValidateServerName_ErrorMessages` — error messages include actionable guidance
- [x] `gofmt` clean, `go vet` clean, full package test suite passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0183ha4LnfrgzgTuEWMywUw1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -391,6 +391,7 @@ The `mcp-grafana` binary supports various command-line flags for configuration:
 - `--address`: The host and port for SSE/streamable-http server - default: `localhost:8000`
 - `--base-path`: Base path for the SSE/streamable-http server
 - `--endpoint-path`: Endpoint path for the streamable-http server - default: `/`
+- `--server-name`: Server name used in the MCP handshake and OTel `service.name` - default: `grafana-mcp-server`. Overrides `GRAFANA_MCP_SERVER_NAME` env var
```
Comment (ioanarm) [inline README.md]: The compiled binary reports `--server-name` default mcp-grafana (const defaultServerName = "mcp-grafana"), but this line says grafana-mcp-server. Since this is the operator-facing contract, can you updated it?

File: cmd/mcp-grafana/main.go [hunk 2]
```diff
@@ -659,6 +689,22 @@ func main() {
 		os.Exit(2)
 	}
 
+	serverNameFlagSet := false
+	flag.Visit(func(f *flag.Flag) {
+		if f.Name == "server-name" {
+			serverNameFlagSet = true
+		}
+	})
+	serverName = resolveServerName(serverName, serverNameFlagSet, os.Getenv("GRAFANA_MCP_SERVER_NAME"), defaultServerName)
```
Comment (ioanarm) [inline cmd/mcp-grafana/main.go:813]: Every existing env var in the repo uses the `GRAFANA_` prefix directly (`GRAFANA_URL`, `GRAFANA_ORG_ID`, `GRAFANA_SERVICE_ACCOUNT_TOKEN`, `GRAFANA_API_KEY`). 
This introduces a new `GRAFANA_MCP_` prefix. Was that intentional to disambiguate "the MCP server's name" from Grafana itself? If there's no strong reason, `GRAFANA_SERVER_NAME` would match the established convention, though I can see the argument that `GRAFANA_MCP_SERVER_NAME` reads clearer. WDYT?

**Review by ioanarm — COMMENTED** [review]
Thanks for this 🚀 ! A couple of comments from my side!

File: README.md (hunk 1 above)
Comment (austinmilt) [reply README.md]: oof, nice catch! Will definitely update.

File: cmd/mcp-grafana/main.go (hunk 2 above)
Comment (austinmilt) [reply cmd/mcp-grafana/main.go:813]: A lot of the env variables refer to a running grafana instance, whereas this refers specifically to the MCP server name. Keeping the `_MCP` makes it clear what server it's referring to. If that's not in line with the rest of the repo though I am happy to change it.

Comment (ioanarm) [reply cmd/mcp-grafana/main.go:813]: Happy to leave it as is. Thanks!

**Review by ioanarm — DISMISSED** [review]

**Comment by ioanarm** [comment]
In order to merge it, commits need to be verified. Sorry for the extra effort on this one.

**Comment by gerboland** [comment]
Conflict in README.md, otherwise I'm 100% happy to approve

**Comment by austinmilt** [comment]
@gerboland is this still good to go?

**Comment by ioanarm** [comment]
@austinmilt sorry there is another conflict on a test :(! Can you please resolve this one as well? 
It's good to go!

**Comment by austinmilt** [comment]
> @austinmilt sorry there is another conflict on a test :(! Can you please resolve this one as well? It's good to go!

@ioanarm done o7

**Review by ioanarm — APPROVED** [review]
